[RF-DOCS][ci skip] Update Active Record Encryption Guide (#55188)
rails/rails: guides/source/active_record_encryption.md

Active Record Encryption
========================

This guide covers how to encrypt data in your database using Active Record.

After reading this guide, you will know:

* How to set up database encryption with Active Record.
* How to migrate unencrypted data.
* How to make different encryption schemes coexist.
* More about advanced concepts like Encryption Contexts and Key Providers.

--------------------------------------------------------------------------------

> Active Record Encryption exists to protect sensitive information in your application, such as personally identifiable information (PII) about your users. Active Record supports application-level encryption by allowing you to declare which attributes should be encrypted. It enables transparent encryption and decryption of attributes when saving and retrieving data. The encryption layer sits between the application and the database.

Author: Removed.

way, meaning it will produce the same encrypted output given the same input
value. This makes querying encrypted attributes possible, like the `email`
above.

The `:deterministic` option allows for querying by trading off lesser security.
The data is still encrypted but the determinism makes crypto-analysis easier.
For this reason, non-deterministic encryption is recommended for all data unless
you need to query by an attribute.

NOTE: In non-deterministic mode, Active Record uses
[AES](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard)-[GCM](https://en.wikipedia.org/wiki/Galois/Counter_Mode)
with a 256-bits key and a random initialization vector. In deterministic mode,
it also uses AES-GCM, but the initialization vector is not random. It is generated as an a function of the key and the plaintext content ([HMAC](https://en.wikipedia.org/wiki/HMAC)-SHA-256 digest of the two).

NOTE: If you do not define a `deterministic_key`, then you have effectively disabled deterministic encryption.

## Basic Usage

### Ignoring Case

You might need to ignore the case when querying deterministically encrypted data. Two approaches make accomplishing this easier:

> You can use the `:downcase` option when declaring the encrypted attribute to downcase the content before encryption occurs.

Reviewer: Nice! This section reads a lot better now.

```ruby
class Person
  encrypts :email_address, deterministic: true, downcase: true
end
```

When using `:downcase`, the original case is lost. In some situations, you might
want to ignore the case only when querying while also storing the original case.
For those situations, you can use the `:ignore_case` option. This requires you
to add a new column named `original_<column_name>` to store the content with the
case unchanged:

```ruby
class Label
  encrypts :name, deterministic: true, ignore_case: true # the content with the original case will be stored in the column `original_name`
end
```

### Serialized Attributes

By default, Active Record Encryption will serialize values using the underlying type before encrypting them as long as the value is serializable as Strings. If the underlying type is not serializable as a String, you can use a custom [`message_serializer`](https://edgeapi.rubyonrails.org/classes/ActiveRecord/Encryption/MessageSerializer.html):

```ruby
class Article < ApplicationRecord
  encrypts :metadata, message_serializer: SomeCustomMessageSerializer.new
end
```

> Attributes with structured types using the [`serialized`](https://api.rubyonrails.org/v8.0.2/classes/ActiveRecord/AttributeMethods/Serialization/ClassMethods.html#method-i-serialize) method are supported with encryption as well. The `serialized` method is used when you have an attribute that needs to be saved to the database as a serialized object (using `YAML`, `JSON` or such), and retrieved by deserializing into the same object.

Reviewer: Nitpick.
```suggestion
Attributes with structured types using the [`serialized`](https://api.rubyonrails.org/v8.0.2/classes/ActiveRecord/AttributeMethods/Serialization/ClassMethods.html#method-i-serialize) method support encryption as well. The `serialized` method is used when you have an attribute that needs to be saved to the database as a serialized object (using `YAML`, `JSON` or such), and retrieved by deserializing into the same object.
```

...or maybe:
```suggestion
Attributes with structured types using the [`serialized`](https://api.rubyonrails.org/v8.0.2/classes/ActiveRecord/AttributeMethods/Serialization/ClassMethods.html#method-i-serialize) method can be encrypted as well. The `serialized` method is used when you have an attribute that needs to be saved to the database as a serialized object (using `YAML`, `JSON` or such), and retrieved by deserializing into the same object.
```

end
```

> ### Unique Constraints with Encrypted Data

Reviewer: I was expecting this section to describe database constraints.
Maybe avoid that term?
```suggestion
### Ensuring Uniqueness with Encrypted Data
```

Unique constraints are only supported with deterministically encrypted data.

#### Unique Validations

In order to support unique validations, you'll need to enable extended queries. The default value for this configuration is false:

```ruby
# config/applicaiton.rb
config.active_record.encryption.extend_queries = true
```

Then, the uniqueness constraint can be specified normally, along with encryption:

```ruby
class Person
  validates :email_address, uniqueness: true
  encrypts :email_address, deterministic: true, downcase: true
end
```

The `extended_queries` configuration also allows combining encrypted and unencrypted data, and when configuring previous encryption schemes.

> NOTE: If you want to ignore the case, make sure to use the `downcase` or `ignore_case` option in the `encrypts` declaration. Using the `case_sensitive` option in the validation won't work.

Reviewer: It was a bit unclear to me this note is for uniqueness.
I think it's also clearer to use colons as a prefix for options here:
Maybe something like:
```suggestion
NOTE: If you want to ignore the case for uniqueness, make sure to use the `:downcase` or `:ignore_case` option in the `encrypts` declaration. Using the `:case_sensitive` option in the validation won't work.
```

end
```

### Ensuring Uniqueness with Encrypted Data

Checking for uniqueness is only supported with deterministically encrypted data.

#### Unique Validations

> In order to support unique validations, you'll need to enable extended queries.
> The default value for this configuration is false:

Reviewer: I think it would be clearer if we can move the reasoning of using `extend_queries` on line 265 here, so there is less jumping of context?
Maybe something like:
```suggestion
In order to support unique validations, you'll need to enable extended queries. This makes sure we query for the downcased encrypted attribute (if used) and previous encryption schemes as well. The default value for this configuration is false:
```

```ruby
# config/applicaiton.rb
config.active_record.encryption.extend_queries = true
```

Then, the uniqueness constraint can be specified normally, along with
encryption:

```ruby
class Person
  validates :email_address, uniqueness: true
  encrypts :email_address, deterministic: true, downcase: true
end
```

> The `extended_queries` configuration also allows combining encrypted and
> unencrypted data, and when configuring previous encryption schemes.

Reviewer: I we move this into the opening paragraph, we can remove it here
```suggestion
```

In case you need to disable filtering of encrypted parameters, you can use the
following configuration:

```ruby
# config/applicaiton.rb
config.active_record.encryption.add_to_filter_parameters = false
```

When filtering is enabled, if you want to exclude specific attributes from
automatic filtering, you can use this configuration:

```ruby
config.active_record.encryption.excluded_from_filter_parameters = [:catchphrase]
```

NOTE: When generating the filter parameter, Rails will use the model name as a
prefix. E.g: For `User#email`, the filter parameter will be `user.email`.

### Action Text

You can encrypt Action Text attributes by passing `encrypted: true` in their
declaration.

```ruby
class Message < ApplicationRecord
  has_rich_text :content, encrypted: true
end
```

NOTE: Passing individual encryption options to Action Text attributes is not
supported. It will use non-deterministic encryption with the global encryption
options configured.

### Fixtures

To allow your tests can use plain text values in the YAML fixture files for encrypted attributes, you can configure fixtures to be automatically encrypted by adding this configuration to your `test.rb` file:

```ruby

> config.active_record.encryption.encrypt_fixtures = true

Reviewer: Maybe make it clear this needs to be set in the `configure` block?
```suggestion
Rails.application.configure do
  ...
  config.active_record.encryption.encrypt_fixtures = true
  ...
end
```

organization's security policy or if you suspect a key may be compromised.

In the example below, the *last key* is used for encrypting new content and all
keys are tried when decrypting content until one works.

```yml
active_record_encryption:
  primary_key:
    - a1cc4d7b9f420e40a337b9e68c5ecec6 # Previous keys can still decrypt existing content
    - bc17e7b413fd4720716a7633027f8cc4 # Active, encrypts new content
  key_derivation_salt: a3226b97b3b2f8372d1fc6d497a0c0d3
```

This enables key rotation workflow where you keep a short list of keys by adding
new keys, re-encrypting content, and deleting old keys.

NOTE: Rotating keys is not supported for deterministic encryption.

### Storing Key References

You can store a reference to the encryption key in the encrypted message itself.
The advantage of doing this is that decryption can be more performant as the
system does not have to try a list of keys to find one that works. The tradeoff
is that the encryption data will be a bit larger.

In order to store a key reference, you need to enable this configuration:

```ruby
config.active_record.encryption.store_key_references = true
```

> ## Configuration
> 
> You can configure Active Record Encryption options in your `application.rb`
> (most common scenario) or in a specific environment config file
> `config/environments/<env name>.rb` if you want to set them on a per-environment
> basis. You can see the Active Record Encryption related configuration properties
> in the [Configuration
> Guide](configuring.html#config-active-record-encryption-support-unencrypted-data)
> 

Reviewer: I think we can remove this section now.
It's pretty generic and other guides don't have a configuration section either .
```suggestion
```

Most users would not need to use them.

The main components of encryption contexts are:

* `encryptor`: exposes the internal API for encrypting and decrypting data.  It
  interacts with a `key_provider` to build encrypted messages and deal with
  their serialization. The encryption/decryption itself is done by the `cipher`
  and the serialization by `message_serializer`.
* `cipher`: the encryption algorithm itself (AES 256 GCM).
* `key_provider`: serves encryption and decryption keys.
* `message_serializer`: serializes and deserializes encrypted payloads.

WARNING: If you decide to build your own `message_serializer`, it's important to
use safe mechanisms that can't deserialize arbitrary objects. A commonly
supported scenario is encrypting existing unencrypted data. An attacker can
leverage this to enter a tampered payload before encryption takes place and
perform RCE attacks. This means custom serializers should avoid `Marshal`,
`YAML.load` (use `YAML.safe_load`  instead), or `JSON.load` (use `JSON.parse`
instead).

### Built-In Encryption Context

The global encryption context is the one used by default and is configured with

> other configuration properties in your `application.rb` or environment config

Reviewer: ```suggestion
other configuration properties in your `config/application.rb` or environment config
```

serves keys derived from the provided passwords using PBKDF2. This is the key
provider configured by default.

```ruby
config.active_record.encryption.key_provider = ActiveRecord::Encryption::DerivedSecretKeyProvider.new(["some passwords", "to derive keys from. ", "These should be in", "credentials"])
```

NOTE: By default, `active_record.encryption` configures a
`DerivedSecretKeyProvider` with the keys defined in
`active_record.encryption.primary_key`.

#### EnvelopeEncryptionKeyProvider

The
[`EnvelopeEncryptionKeyProvider`](https://api.rubyonrails.org/classes/ActiveRecord/Encryption/EnvelopeEncryptionKeyProvider.html)
implements a simple [envelope
encryption](https://en.wikipedia.org/wiki/Hybrid_cryptosystem#Envelope_encryption)
strategy, where the data is encrypted with a key, which in turn is also
encrypted.

The `EnvelopeEncryptionKeyProvider` generates a random key for each data
encryption operation. It stores the data-key with the data itself. Then, the
data-key is also encrypted with a primary key defined in the credential
`active_record.encryption.primary_key`.

You can configure Active Record to use this key provider by adding this to your

> `application.rb`:

Reviewer: ```suggestion
`config/application.rb`:
```
